File: 0145-binary-tree-postorder-traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void display(TreeNode* root, vector<int>& v) // left ORDER DISPLAY
    {
        if (root == nullptr)
            return;
       
        display(root->left,v);
        display(root->right,v);
         v.push_back(root->val);
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> v;
        display(root, v);
        return v;
    }
};

File: 0145-binary-tree-postorder-traversal/solution.java
class Solution {

    public List<Integer> postorderTraversal(TreeNode root) {

        List<Integer> res = new ArrayList<>();

        if (root == null)
            return res;

        Stack<TreeNode> stack = new Stack<>();
        TreeNode node = root;

        while (node != null || !stack.isEmpty()) {

            if (node != null) {
                stack.push(node);
                node = node.left;
            }

           else if (stack.peek().right != null) {
                node = stack.peek().right;
            } else {

                TreeNode temp = stack.pop();
                res.add(temp.val);

                while (!stack.isEmpty() &&
                       temp == stack.peek().right) {

                    temp = stack.pop();
                    res.add(temp.val);
                }
            }
        }

        return res;
    }
}
